internal/wal/wal_aof.go

```diff
@@ -18,43 +20,146 @@ import (
+	}
 
-func (w *WALAOF) Init(t time.Time) error {
-	slog.Debug("initializing WAL at", slog.Any("log-dir", w.logDir))
-	if err := os.MkdirAll(w.logDir, os.ModePerm); err != nil {
-		return fmt.Errorf("failed to create log directory: %w", err)
+	if rotationMode != "time" && rotationMode != "size" {
+		return nil, fmt.Errorf("invalid rotationMode: %s, must be 'time' or 'size'", rotationMode)
+	}
+
+	if rotationMode == "time" && rotationInterval <= 0 {
+		return nil, fmt.Errorf("rotationInterval must be greater than zero when rotationMode is 'time'")
 	}
 
-	timestamp := t.Format("20060102_1504")
-	path := filepath.Join(w.logDir, fmt.Sprintf("wal_%s.aof", timestamp))
+	w := &WALAOF{
+		logDir:              logDir,
+		syncInterval:        time.Millisecond * 100,
+		lastSyncTime:        time.Now(),
+		shouldFsync:         enableFsync,
+		maxFileSize:         maxFileSize,
+		maxSegments:         maxSegments,
+		currentSegmentIndex: 0,
+		rotationMode:        rotationMode,
+		rotationInterval:    rotationInterval,
+		isBuffered:          isBuffered,
+		writeMode:           writeMode,
+		quit:                make(chan struct{}),
+	}
 
-	newFile, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	files, err := filepath.Glob(filepath.Join(logDir, segmentPrefix+"*"))
 	if err != nil {
-		return fmt.Errorf("failed to open new WAL file: %v", err)
+		return nil, err
 	}
 
-	w.file = newFile
-	return nil
+	if len(files) > 0 {
+		lastSegmentID, err := findLastSegmentIndex(files)
+		if err != nil {
+			return nil, err
+		}
+		w.currentSegmentIndex = lastSegmentID
+	}
+
+	if err := w.Init(time.Now()); err != nil {
+		return nil, err
+	}
+
+	if w.rotationMode == "time" {
+		w.rotationTimer = time.NewTimer(w.rotationInterval)
+
+		go w.handleRotationTimer()
+	}
+
+	return w, nil
 }
 
-// LogCommand serializes a WALLogEntry and writes it to the current WAL file.
-func (w *WALAOF) LogCommand(c *cmd.DiceDBCmd) {
+func (w *WALAOF) Init(t time.Time) error {
 	w.mutex.Lock()
 	defer w.mutex.Unlock()
 
+	segmentFileName := fmt.Sprintf("%s%d", segmentPrefix, w.currentSegmentIndex)
+	path := filepath.Join(w.logDir, segmentFileName)
+
+	file, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to open WAL file: %v", err)
+	}
+
+	w.file = file
+
+	if w.isBuffered {
+		w.bufWriter = bufio.NewWriterSize(w.file, 4096) // Use a buffer size of 4KB
+		w.writer = w.bufWriter
+	} else {
+		w.writer = w.file
+	}
+
+	return nil
+}
+
+func (w *WALAOF) handleRotationTimer() {
+	for {
+		select {
+		case <-w.rotationTimer.C:
+			w.mutex.Lock()
+			err := w.rotateLog()
+			if err != nil {
+				slog.Warn("failed to rotate WAL log on timer", slog.Any("error", err.Error()))
+			}
+			w.mutex.Unlock()
+		case <-w.quit:
+			return
+		}
+	}
+}
+
+// LogCommand serializes a WALLogEntry and writes it to the current WAL segment.
+func (w *WALAOF) LogCommand(c *cmd.DiceDBCmd) {
```

Comment:
Usually the wal does not understand the data which it is writing. We could probably modify this method to only accept a byte array.